### PR TITLE
PSFR-1896 - Create new end point for Assign and Unassign

### PR DESCRIPTION
**Pull request by abdulrkalam**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/CaseAllocationService.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    val caseList = emptyList<CaseAllocationEntity?>().toMutableList()
+    if (caseAllocation.staffId != null && caseAllocation.staffFirstName != null && caseAllocation.staffLastName != null) {
+      for (nomsId in caseAllocation.nomsIds) {
+        val prisoner = prisonerRepository.findByNomsId(nomsId)
+          ?: throw ResourceNotFoundException("Prisoner with id $nomsId not found in database")
+        val caseAllocationExists = caseAllocationRepository.findByPrisonerIdAndIsDeleted(
+          prisoner.id(),
+          false,
+        )
+        if (caseAllocationExists != null) {
```
Comment (mkerry-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/CaseAllocationService.kt:77]: when we assign a case that's already assigned we'd expect it just to be assigned to the new person and the old record be deleted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/data/CaseAllocation.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data
+
+data class CaseAllocation(
```
Comment (mkerry-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/data/CaseAllocation.kt:3]: this is only used in tests, should probably be in src/test/

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/CaseAllocationService.kt (hunk 1 above)
Comment (abdulrkalam) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/CaseAllocationService.kt:76]: Done

File: src/main/resources/db/migration/V1_65__add_case_allocation.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+(
+    id                              serial primary key,
+    prisoner_id                     integer not null references prisoner (id),
+    staff_id                        integer not null,
+    staff_firstname                 varchar(100) not null,
+    staff_lastname                  varchar(100) not null,
+    is_deleted                      boolean not null default false,
+    when_created                    timestamp with time zone not null default now(),
+    deleted_at                      timestamp with time zone
+);
```
[the remark this answers is not in the document]
Comment (abdulrkalam) [reply src/main/resources/db/migration/V1_65__add_case_allocation.sql:11]: As we do soft delete, we cannot have unique index. Created the index for prisoser_id and staff_id

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/data/CaseAllocation.kt (hunk 2 above)
Comment (abdulrkalam) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/data/CaseAllocation.kt:3]: This is also used in POST as well, so not moving to test

**Pull request merged**